cmd/metal-api/internal/service/machine-service.go

```diff
@@ -1516,66 +1516,74 @@ func (r machineResource) freeMachine(request *restful.Request, response *restful
-	if checkError(request, response, utils.CurrentFuncName(), err) {
-		return
+	sw, err := setVrfAtSwitches(ds, m, "")
+	logger.Infow("set VRF at switch", "machineID", m.ID, "error", err)
+	if err != nil {
+		return err
 	}
 
 	deleteEvent := metal.MachineEvent{Type: metal.DELETE, Old: m}
-	err = r.Publish(metal.TopicMachine.GetFQN(m.PartitionID), deleteEvent)
-	utils.Logger(request).Sugar().Infow("published machine delete event", "machineID", id, "error", err)
-	if checkError(request, response, utils.CurrentFuncName(), err) {
-		return
+	err = publisher.Publish(metal.TopicMachine.GetFQN(m.PartitionID), deleteEvent)
+	logger.Infow("published machine delete event", "machineID", m.ID, "error", err)
+	if err != nil {
+		return err
 	}
 
 	if m.Allocation != nil {
 		// if the machine is allocated, we free it in our database
-		err = r.releaseMachineNetworks(m, m.Allocation.MachineNetworks)
+		err = releaseMachineNetworks(ds, ipamer, m, m.Allocation.MachineNetworks)
 		if err != nil {
 			// TODO: Trigger network garbage collection
 			// TODO: Check if all IPs in rethinkdb are in the IPAM and vice versa, cleanup if this is not the case
 			// TODO: Check if there are network prefixes in the IPAM that are not in any of our networks
-			utils.Logger(request).Sugar().Errorf("an error during releasing machine networks occurred, scheduled network garbage collection", "error", err)
+			logger.Errorw("an error during releasing machine networks occurred, scheduled network garbage collection", "error", err)
 		}
 	}
 
 	old := *m
 	m.Allocation = nil
 	m.Tags = nil
-	err = r.ds.UpdateMachine(&old, m)
-	if checkError(request, response, utils.CurrentFuncName(), err) {
-		return
+	err = ds.UpdateMachine(&old, m)
+	if err != nil {
+		return err
 	}
-	utils.Logger(request).Sugar().Infow("freed machine", "machineID", id)
+	logger.Infow("freed machine", "machineID", m.ID)
 
 	switchEvent := metal.SwitchEvent{Type: metal.UPDATE, Machine: *m, Switches: sw}
-	err = r.Publish(metal.TopicSwitch.GetFQN(m.PartitionID), switchEvent)
-	utils.Logger(request).Sugar().Infow("published switch update event", "machineID", id, "error", err)
-	if checkError(request, response, utils.CurrentFuncName(), err) {
-		return
-	}
-	err = response.WriteHeaderAndEntity(http.StatusOK, makeMachineResponse(m, r.ds, utils.Logger(request).Sugar()))
+	err = publisher.Publish(metal.TopicSwitch.GetFQN(m.PartitionID), switchEvent)
+	logger.Infow("published switch update event", "machineID", m.ID, "error", err)
 	if err != nil {
-		zapup.MustRootLogger().Error("Failed to send response", zap.Error(err))
-		return
+		return err
 	}
+
+	return nil
 }
 
-func (r machineResource) releaseMachineNetworks(machine *metal.Machine, machineNetworks []*metal.MachineNetwork) error {
+func releaseMachineNetworks(ds *datastore.RethinkStore, ipamer ipam.IPAMer, machine *metal.Machine, machineNetworks []*metal.MachineNetwork) error {
```

Comment:
We can ommit the last function parameter, because machineNetworks is part of the machine struct.